locator/types.hh

```diff
@@ -21,12 +25,101 @@ namespace locator {
 
 using inet_address = gms::inet_address;
 
+class token_metadata;
+class topology;
+
+struct host_id_or_endpoint {
+    host_id id;
+    gms::inet_address endpoint;
+
+    enum class param_type {
+        host_id,
+        endpoint,
+        auto_detect
+    };
+
+    host_id_or_endpoint(const sstring& s, param_type restrict = param_type::auto_detect);
+
+    bool has_host_id() const noexcept {
+        return bool(id);
+    }
+
+    bool has_endpoint() const noexcept {
+        return endpoint != gms::inet_address();
+    }
+
+    // Map the host_id to endpoint based on whichever of them is set,
+    // using the token_metadata
+    void resolve(const token_metadata& tm);
+};
+
 // Endpoint Data Center and Rack names
 struct endpoint_dc_rack {
     sstring dc;
     sstring rack;
+
+    bool operator==(const endpoint_dc_rack& o) const noexcept {
+        return dc == o.dc && rack == o.rack;
+    }
+    bool operator!=(const endpoint_dc_rack& o) const noexcept {
+        return !(*this == o);
+    }
 };
 
 using dc_rack_fn = seastar::noncopyable_function<endpoint_dc_rack(inet_address)>;
 
+class node : public enable_lw_shared_from_this<node> {
+public:
+    using pending = bool_class<struct pending_tag>;
+
+private:
+    locator::host_id _host_id;
+    inet_address _endpoint;
+    endpoint_dc_rack _dc_rack;
+    pending _is_pending;
+
+    using node_member_hook = boost::intrusive::list_member_hook<boost::intrusive::link_mode<boost::intrusive::auto_unlink>>;
+
+    node_member_hook _nodes_link;
+    node_member_hook _dc_nodes_link;
+    node_member_hook _rack_nodes_link;
+
```

Comment:
I guess we can have state that is shared among different iterations of a node in different topology objects, and state that is local to a node in different topology objects. For example, host_id is shared, and IP address is private. Dead/alive state is up to is to decide (if we want to precompute quorums for degraded clusters).

Reply:
I see where you're going, it will make copying nodes across shards harder, like I propose to do when "broadcasting" the topology to all shards after it's mutated on shard 0, since we'll to maintain the sharing relationship respectively on each shard.  For example:
- shard 0:
  - topology 0: node{X}[s0, t0] -> shared_node_state{X}[s0]
  - topology 1: node{X}[s0, t1] -> shared_node_state{X}[s0]
- shard I:
  - topology 0: node{X}[sI, t0] -> shared_node_state{X}[sI]

Now when we want to clone node{X}[s0, t1] to node{X}[sI, t1]
we should end up with:
  - topology 1: node{X}[sI, t1] -> shared_node_state{X}[sI]

so the cloning process need to locate shared_node_state{X} on the destination shard
and tie up the ends correctly.

Reply:
Yes. One way out is to use node indexes instead of node pointers. It's actually justified since this is a heavyweight structure and node pointers are a large component of it, but it shouldn't be dominating this change (the optimization is a nice add on but shouldn't derail us).